models/Element/Recyclebin/Item.php

```diff
@@ -329,6 +326,67 @@ protected function doRecursiveRestore(Element\ElementInterface $element)
         }
     }
 
+    /**
+     * @param Element\ElementInterface $data
+     *
+     * @return mixed
+     */
+    public function marshalData($data)
+    {
+        //for full dump of relation fields in container types
+        $copier = new DeepCopy();
+        $copier->addTypeFilter(
+            new \DeepCopy\TypeFilter\ReplaceFilter(
+                function ($currentValue) {
+                    $elementType = Element\Service::getType($currentValue);
+                    $descriptor = new Model\Version\ElementDescriptor($elementType, $currentValue->getId());
+                    return $descriptor;
+                }
+            ),
+            new class($this->element) extends TypeMatcher {
+                /**
+                 * @param mixed $element
+                 *
+                 * @return boolean
+                 */
+                public function matches($element)
+                {
+                    //compress only elements with full_dump_state = false
+                    return $element instanceof Element\ElementInterface && $element instanceof Element\ElementDumpStateInterface && !($element->isInDumpState());
+                }
+            }
+        );
+        $copier->addFilter(new Model\Version\SetDumpStateFilter(true), new \DeepCopy\Matcher\PropertyMatcher(Element\ElementDumpStateInterface::class, Element\ElementDumpStateInterface::DUMP_STATE_PROPERTY_NAME));
```

Comment:
Please check in which order the filters get applied. SetDumpStateFilter must be executed first I guess.